Add more Item sorting | Add Radiobackpack Category

Lazejun opened the pull request:
## What type of PR is this.
1. [ ] Bug
2. [x] Change
3. [x] Enhancement

### What have you changed and why?
Information:
**Why?:**

The less unusable items that spawn in loot the higher the chance to get usefull items.

Added a new Cateogory of Backpacks for TFAR-Radios/Radios that are Compatable with it, these are Filtered for Radio Encryption depending on Teamplayer.

Added allBackpackRadio to LootDevice

They were Defined in **allBackpacksEmpty** before.
Added Side Filtering for BackpacksRadio, only Green/Blue Side compatable will be given
**allBackpackRadio**
--------> The filtering only takes out Radios for Rebell Side, other Radiobackpacks  remain in Allbackpacksempty

Added Side Filtering for Item allRadios ------> Only Radios of your Side apear in Loot.
**allRadios**

Changed "rhsusf_Rhino" to Unknown Items in Category Overrides

Added Side filtering for **allUAVTerminals**.
Only Green and Blue given depending on Side.

Also noticed 1 CSW Weaponpart still being listed in Lootweapon
-------> Added that Weaponpart in EquipmentOverrides.

### Please specify which Issue this PR Resolves.
closes #731 

### Please verify the following and ensure all checks are completed.

1. [x] Have you loaded the mission in singleplayer?
2. [x] Have you loaded the mission in LAN host?
3. [ ] Have you loaded the mission on a dedicated server?

### Is further testing or are further changes required?
1. [x] No
2. [ ] Yes (Please provide further detail below.)

### How can the changes be tested?

Steps: All Modsets:
UAVTerminal Filtering:
Get ingame 
```sqf
allUAVTerminals
``` 
It should return only 1 Entry which is either "B_UavTerminal" for Bluefor and "I_UavTerminal" for Greenfor. (I_E_UavTermianl if Contact enable on Green side)

Radio Filtering:
With TFAR
```sqf
allRadios
```
It should only return "tf_anprc148jem" for Green and "tf_anprc152" for Blue.

Radiobackpack sorting:
With TFAR:
```sqf
allBackpacksRadio
```
and
```sqf
lootDevice
```
It should only Return side fitting Equipment, Device filtering worked before, i only added Radiobackpacks to it, to also Increase the Chances for them to spawn. (The Parameter is also 2 by Default).
********************************************************
Notes: Having BackpacksRadio actually apear in Loot by default would make #720 being turned off by default more Logical as when it first came around.

Since lootBackpack USED to Contain the Radiobackpacks but Default Value for them is 0 so they dont spawn.

In Theory its Future Proof, cause it would detect Radios for TFAR and TFAR Beta.

Examples of Allbackpacksradio with only TFAR Loaded:

BlueSide:
```sqf
["tf_rt1523g","tf_rt1523g_bwmod","tf_rt1523g_rhs","tf_rt1523g_big","tf_rt1523g_big_bwmod","tf_rt1523g_big_bwmod_tropen","tf_rt1523g_big_rhs","tf_rt1523g_sage","tf_rt1523g_green","tf_rt1523g_fabric","tf_rt1523g_black","tf_anarc210"]
```

Green Side:
```sqf
["tf_anprc155","tf_anprc155_coyote","tf_anarc164"]
```

BlueSide WILL Explode with 3CB on Blue Side due to ALL of the British Backpack-Radio Variants.

jaj22 commented on A3-Antistasi/functions/Ammunition/fn_equipmentSort.sqf:
```diff
@@ -137,3 +137,85 @@ allCivilianGlasses deleteAt (allCivilianGlasses find "G_I_Diving");
+//////////////////
+//Removes Radios that are not useable and adds filtering for Sides.(TFAR)
+if (HasTFAR) then {
+allRadios deleteAt (allRadios find "ItemRadio");
+};
+allRadios deleteAt (allRadios find "tf_pnr1000a"); //Redfor 2K Radio
+allRadios deleteAt (allRadios find "tf_fadak"); //Redfor 5K Radio
+allRadios deleteAt (allRadios find "tf_rf7800str"); //Bluefor 2K Radio
+allRadios deleteAt (allRadios find "tf_anprc154"); //Greenfor 2K Radio
+if (teamPlayer isEqualTo west) then {
+allRadios deleteAt (allRadios find "tf_anprc148jem");
+} else {
+allRadios deleteAt (allRadios find "tf_anprc152");
+};
```
So this is a bit interesting. Removing items from the radio category when they are, in fact, radios. I guess it's valid if the categories are purely for rebel purposes and these items aren't usable for the defined purpose. I do worry a bit that we might need categories for other purposes in the future.

If the purpose is just to get them out of crates, the soft option would be to create a tfarModCompat in functions/Templates, and delete this stuff from lootItem/lootDevice there. Not sure what the right option is at the moment though.

jaj22 commented on A3-Antistasi/functions/Ammunition/fn_equipmentSort.sqf:
```diff
@@ -137,3 +137,85 @@ allCivilianGlasses deleteAt (allCivilianGlasses find "G_I_Diving");
+//   UAVTerminal   ///
+//////////////////
+//Removes Terminals that are not useable and adds filtering for Sides.
+allUAVTerminals deleteAt (allUAVTerminals find "O_UavTerminal"); //Redfor Terminal
+allUAVTerminals deleteAt (allUAVTerminals find "C_UavTerminal"); //Civilian
+allUAVTerminals deleteAt (allUAVTerminals find "I_E_UavTerminal"); //LDF Terminal
+if (teamPlayer isEqualTo west) then {
+allUAVTerminals deleteAt (allUAVTerminals find "I_UavTerminal");
+} else {
+allUAVTerminals deleteAt (allUAVTerminals find "B_UavTerminal");
```
These appear to get their side locking from itemInfo >> side, although maybe there's some hardcoded stuff underneath, especially if the LDF version doesn't work the same as the standard greenfor terminal. Short list anyway, and probably won't change, so it doesn't much matter.

Similar point about whether you want to delete these from the category or just from loot though.

jaj22 reviewed (commented):
Submitted as comment because I don't really know what I'm doing.

Lazejun commented:
Ok Updated the PR with your Suggestions, Used the Category Overrides to remove the Rhino Headmount instead of removing it from the other Categories.

With the LDF-UAV Terminal i came up with the Solution of i dont care, it is ofcourse the same but Idea was so Blue and Green both only have 1 UAV terminal which would make the chance to Unlock sort of "Fairer" in each Game but its only if Enoch is enabled in the end it doesent matter that much.

Removed the 3CB Faction Radios from Cateogory overrides due to them not working with TFAR anymore, 3CB BAF Radios work fine.

jaj22 reviewed (approved):
Looks good.

Lazejun commented:
So i was curious how well the Radio Filtering would work with TFAR BETA and found out that it changed:
"tf_guer_radio_code"
 to 
"tf_independent_radio_code"

And also leaving (_encrypt isEqualTo "") in leads to Issues as it still giving Radios that have no Encryption set so i removed that aswell.

I also got rid of Equipment overrides for the BackpackRadios and let used the Filtering for Encryption to filter from AllBackpacksEmpty.
The only new Issue would be that now Unfiltered/Left alone Radios remain in BackpacksEmpty.

I will update the PRs Description in a bit with the New stuff, cause most is now taken care by Autodetection.

jaj22 commented on A3-Antistasi/functions/Ammunition/fn_equipmentSort.sqf:
```diff
@@ -141,19 +141,19 @@ allCivilianGlasses deleteAt (allCivilianGlasses find "LIB_Glasses");
 //   Radios   ///
 //////////////////
-private _encryptRebel = if (teamPlayer == west) then { "tf_west_radio_code" } else { "tf_guer_radio_code" };
+private _encryptRebel = if (teamPlayer == west) (teamPlayer == west) then { ["tf_west_radio_code"] } else { ["tf_guer_radio_code" or "tf_independent_radio_code"]; };
```
Looks like the condition got doubled up, and the array at the end should have a comma in the middle rather than an `or`, unless I'm missing something.

jaj22 replied:
Never mind, already fixed.

jaj22 commented on A3-Antistasi/functions/Ammunition/fn_equipmentSort.sqf:
```diff
@@ -141,19 +141,19 @@ allCivilianGlasses deleteAt (allCivilianGlasses find "LIB_Glasses");
 //   BackpackRadio   ///
 //////////////////
-private _encryptRebel = if (teamPlayer == west) then { "tf_west_radio_code" } else { "tf_guer_radio_code" };
+private _encryptRebel = if (teamPlayer == west) then { ["tf_west_radio_code"] } else { ["tf_guer_radio_code" or "tf_independent_radio_code"] };
```
Broken array here too.

jaj22 commented on A3-Antistasi/functions/init/fn_initVarCommon.sqf:
```diff
@@ -65,7 +65,7 @@ aggregateCategories = ["Weapons", "Items", "Magazines", "Explosives"];
 //It's recommended that these categories be used with caution.
 specialCategories = ["AA", "AT", "GrenadeLaunchers", "LightAttachments", "LaserAttachments", "Chemlights", "SmokeGrenades", "LaunchedSmokeGrenades", "LaunchedFlares", "HandFlares", "IRGrenades","LaserBatteries",
-	"RebelUniforms", "CivilianUniforms", "BackpacksEmpty", "BackpacksTool", "BackpacksStatic", "BackpacksDevice", "CivilianVests", "ArmoredVests", "ArmoredHeadgear", "CivilianHeadgear",
+	"RebelUniforms", "CivilianUniforms", "BackpacksEmpty", "BackpacksTool", "BackpacksStatic", "BackpacksDevice", "BackpackRadio", "CivilianVests", "ArmoredVests", "ArmoredHeadgear", "CivilianHeadgear",
```
Should probably be BackpacksRadio rather than BackpackRadio, to match the others.

Lazejun replied:
Yeah it would make sense for Consistency.

jaj22 commented on A3-Antistasi/functions/Ammunition/fn_equipmentSort.sqf:
```diff
@@ -137,3 +137,33 @@ allCivilianGlasses deleteAt (allCivilianGlasses find "G_I_Diving");
+};
+
+private _encrypthostile = if (teamPlayer == west) then { ["tf_guer_radio_code", "tf_independent_radio_code", "tf_east_radio_code"] } else { ["tf_west_radio_code", "tf_east_radio_code"] };
+allBackpackRadio = [];
```
Already done by initVarServer line 185.

jaj22 commented on A3-Antistasi/functions/Ammunition/fn_equipmentSort.sqf:
```diff
@@ -137,3 +137,33 @@ allCivilianGlasses deleteAt (allCivilianGlasses find "G_I_Diving");
+//   Radios   //
+////////////////
+private _encryptRebel = if (teamPlayer == west) then { ["tf_west_radio_code"] } else { ["tf_guer_radio_code", "tf_independent_radio_code"] };
+allRadios = allRadios select {
+    private _encrypt = getText (configFile >> "CfgWeapons" >> _x >> "tf_encryptionCode");
+    (_encrypt in _encryptRebel);
+};
```
I think allRadios matters slightly for vanilla too, so this filter shouldn't be applied unless tfar is running.

Lazejun replied:
I was actually thinking about that, cause i noticed it Filtered "ItemRadio" away without TFAR, but i wasn't sure if AllRadios was used anywhere else than Loot, cause we check directly for "ItemRadio".

The has TFAR Check wouldnt hurt.

Lazejun commented:
Updated the PR, Changed Backpack to Backpacks, Removed the unneeded empty Array, 
Added the TFAR check which is funny cause apparently our (HasTFAR) uses a Variable that's Common on Both TFARs.

jaj22 reviewed (approved):
All looks good now.
